File: Cpp-C/Ema/Src/Examples/100_Series_Examples/151__Login__AcceptDeny/IProvider.cpp
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright Thomson Reuters 2015. All rights reserved.            --
///*|-----------------------------------------------------------------------------

#include "IProvider.h"

/*
assumptions:
- dependence on the standard / default configuration; e.g.
  - no EmaConfig.xml file
  - standard RDM file dictionary located in the same directory as the app
  - directory requests processed by the EMA
  - dictionary requests processed by the EMA

this interactive provider shows the following functionality
- accepting login requests using the domain Login helper class
- denying login request if not user name
- all other domains are status closed
*/

using namespace thomsonreuters::ema::access;
using namespace thomsonreuters::ema::rdm;
using namespace thomsonreuters::ema::domain::login;
using namespace std;

void AppClient::processLoginRequest( const ReqMsg& reqMsg, const OmmProviderEvent& event )
{
	if ( reqMsg.getNameType() != USER_NAME || reqMsg.getName() != "user" )
	{
		event.getProvider().submit(StatusMsg().domainType(MMT_LOGIN).name(reqMsg.getName()).nameType(USER_NAME)
			.state( OmmState::ClosedEnum, OmmState::SuspectEnum, OmmState::StatusCode::NotAuthorizedEnum, "Login denied" ),
			event.getHandle() );
	}
	else
	{
		RefreshMsg loginRefresh;
		Login::RefreshAttrib refreshAttrib;
		
		Login::ReqAttrib reqAttrib( reqMsg.getAttrib().getElementList() );

		if ( reqAttrib.hasAllowSuspectData() )
			refreshAttrib.allowSuspectData( reqAttrib.getAllowSuspectData() );
		
		if ( reqAttrib.hasSingleOpen() )
			refreshAttrib.singleOpen( reqAttrib.getSingleOpen() );
		
		if ( reqAttrib.hasPosition() )
			refreshAttrib.position( reqAttrib.getPosition() );
		
		if ( reqAttrib.hasApplicationId() )
			refreshAttrib.applicationId( reqAttrib.getApplicationId() );

		event.getProvider().submit( loginRefresh.domainType( MMT_LOGIN ).name( reqMsg.getName() ).nameType( USER_NAME )
			.attrib(refreshAttrib.getData()).solicited(true).complete()
			.state( OmmState::OpenEnum, OmmState::OkEnum, OmmState::NoneEnum, "Login accepted" ),
			event.getHandle() );
	}
}

void AppClient::processInvalidDomainRequest( const ReqMsg& reqMsg, const OmmProviderEvent& event )
{
	event.getProvider().submit( StatusMsg().name( reqMsg.getName() ).serviceName( reqMsg.getServiceName() )
		.domainType( reqMsg.getDomainType() )
		.state( OmmState::ClosedEnum, OmmState::SuspectEnum, OmmState::NotFoundEnum, "Invalid domain" ),
		event.getHandle() );
}

void AppClient::onReqMsg( const ReqMsg& reqMsg, const OmmProviderEvent& event )
{
	switch ( reqMsg.getDomainType() )
	{
	case MMT_LOGIN :
		processLoginRequest( reqMsg, event );
		break;
	default :
		processInvalidDomainRequest( reqMsg, event );
		break;
	}
}

int main( int argc, char* argv[] )
{
	try
	{
		AppClient appClient;

		OmmProvider provider( OmmIProviderConfig().port( "14002" ), appClient );

		sleep( 60000 );
	}
	catch ( const OmmException& excp )
	{
		cout << excp << endl;
	}

	return 0;
}

File: Java/Ema/Src/Examples/com/thomsonreuters/ema/examples/training/iprovider/series100/example151__Login__AcceptDeny/IProvider.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright Thomson Reuters 2016. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.examples.training.iprovider.series100.example151__Login__AcceptDeny;

/*
assumptions:
- dependence on the standard / default configuration; e.g.
  - no EmaConfig.xml file
  - standard RDM file dictionary located in the same directory as the app
  - directory requests processed by the EMA
  - dictionary requests processed by the EMA

this interactive provider shows the following functionality
- accepting login requests using the domain Login helper class
- denying login request if not user name
- all other domains are status closed
*/

import com.thomsonreuters.ema.access.EmaFactory;
import com.thomsonreuters.ema.access.GenericMsg;
import com.thomsonreuters.ema.access.Msg;
import com.thomsonreuters.ema.access.OmmException;
import com.thomsonreuters.ema.access.OmmIProviderConfig;
import com.thomsonreuters.ema.access.OmmProvider;
import com.thomsonreuters.ema.access.OmmProviderClient;
import com.thomsonreuters.ema.access.OmmProviderEvent;
import com.thomsonreuters.ema.access.OmmState;
import com.thomsonreuters.ema.access.PostMsg;
import com.thomsonreuters.ema.access.RefreshMsg;
import com.thomsonreuters.ema.access.ReqMsg;
import com.thomsonreuters.ema.access.StatusMsg;
import com.thomsonreuters.ema.domain.DomainFactory;
import com.thomsonreuters.ema.domain.login.Log;
import com.thomsonreuters.ema.rdm.EmaRdm;


class AppClient implements OmmProviderClient
{	
	public void onReqMsg(ReqMsg reqMsg, OmmProviderEvent event)
	{
		switch (reqMsg.domainType())
		{
			case EmaRdm.MMT_LOGIN :
				processLoginRequest(reqMsg, event);
				break;
			default :
				processInvalidDomainRequest(reqMsg, event);
				break;
		}
	}
	
	public void onRefreshMsg(RefreshMsg refreshMsg,	OmmProviderEvent event){}
	public void onStatusMsg(StatusMsg statusMsg, OmmProviderEvent event){}
	public void onGenericMsg(GenericMsg genericMsg, OmmProviderEvent event){}
	public void onPostMsg(PostMsg postMsg, OmmProviderEvent event){}
	public void onReissue(ReqMsg reqMsg, OmmProviderEvent event){}
	public void onClose(ReqMsg reqMsg, OmmProviderEvent event){}
	public void onAllMsg(Msg msg, OmmProviderEvent event){}
	
	void processLoginRequest(ReqMsg reqMsg, OmmProviderEvent event)
	{
		if ( !reqMsg.name().equals("user") )
		{
			event.provider().submit( EmaFactory.createStatusMsg().domainType(EmaRdm.MMT_LOGIN).name(reqMsg.name()).nameType(reqMsg.nameType())
					.state(OmmState.StreamState.CLOSED, OmmState.DataState.SUSPECT, OmmState.StatusCode.NOT_AUTHORIZED, "Login denied"),
					event.handle() );
		}
		else
		{
			Log.RefreshAttrib refreshAttrib = DomainFactory.Login.createRefreshAttrib();

			Log.ReqAttrib reqAttrib = DomainFactory.Login.createReqAttrib( reqMsg.attrib().elementList() );
			
			if(reqAttrib.hasAllowSuspectData())
				refreshAttrib.allowSuspectData(reqAttrib.allowSuspectData());
			if(reqAttrib.hasSingleOpen())
				refreshAttrib.singleOpen(reqAttrib.singleOpen());

			if(reqAttrib.hasPosition())
				refreshAttrib.position( reqAttrib.position() );
			if(reqAttrib.hasApplicationId())
				refreshAttrib.applicationId( reqAttrib.applicationId() );
			
			event.provider().submit( EmaFactory.createRefreshMsg().domainType(EmaRdm.MMT_LOGIN).name(reqMsg.name()).nameType(EmaRdm.USER_NAME)
					.complete(true).attrib(refreshAttrib.data()).solicited(true)
					.state(OmmState.StreamState.OPEN, OmmState.DataState.OK, OmmState.StatusCode.NONE, "Login accepted"),
					event.handle() );
		}
	}
	
	void processInvalidDomainRequest(ReqMsg reqMsg, OmmProviderEvent event)
	{
		event.provider().submit( EmaFactory.createStatusMsg().name(reqMsg.name()).serviceName(reqMsg.serviceName())
				.state(OmmState.StreamState.CLOSED, OmmState.DataState.SUSPECT,	OmmState.StatusCode.NOT_FOUND, "Invalid domain"),
				event.handle() );
	}
}

public class IProvider
{
	public static void main(String[] args)
	{
		OmmProvider provider = null;
		try
		{	
			AppClient appClient = new AppClient();
			
			OmmIProviderConfig config = EmaFactory.createOmmIProviderConfig();
			
			provider = EmaFactory.createOmmProvider(config.port("14002"), appClient);
			
			Thread.sleep(60000);
		} 
		catch (InterruptedException | OmmException excp)
		{
			System.out.println(excp.getMessage());
		}
		finally 
		{
			if (provider != null) provider.uninitialize();
		}
	}
}
